### ISSUE TITLE
Support DataONE quota checking in Metacat

### ISSUE THREAD

**Issue by csjx** [NCEAS/metacat#1426]
For the DataONE Plus services, we need to be able to enable quota checking and tracking in Metacat.  This needs to be optional, and by default disabled.  We also need to be able to support some quotas, but not all.  For instance, a repository might want to enforce storage quotas, but not portal quotas.

- [ ] Set new properties in `metacat.properties`:
    `dataone.quotas.storage.enabled=false` - Check and report storage quotas
    `dataone.quotas.portals.enabled=false` - Check and report portal quotas
    `dataone.quotas.replication.enabled=false` - Check and report replication quotas

- [ ] Modify the database schema to include a `quotas` table with the following columns:
    - `usage_id` - An autoincremented id
    - `guid` - The identifier of the object from the `systemmetadata` table (`FKEY`)
    - `quota_subject` - The subject of the user/group to report the quota for
    - `quota_id` - The identifier of the quota
    - `quota_name` - The name of the quota ( can be `storage` or `portals`)
    - `date_reported` - The date that the quota usage was reported to the quota service

- [ ] Expose quota property settings in the MN node document

- [ ] Modify Metacat to update the `SystemMetadata.size` to reflect the bytes written (we don't currently verify `size`, but should from now on to prevent under-reporting)

- [ ] Create an `ExecutorService` that accepts `QuotaUsageTask`s that will asynchronously update the Bookkeeper quota usage (see below).

- [ ] Modify `MNodeService` `create()` and `update()` methods to check for storage and portal quotas before completing the request.  See the [example sequence diagram](https://github.com/DataONEorg/bookkeeper/#creating-an-object) as an example of creating a portal document. This will be a call to 

`GET https://api.dataone.org/bookkeeper/v1/quotas/usage?submitter=<submitterSubject>&quotaSubject=<quotaSubject>&quotaName=<quotaName>&requestedUsage=<requestedUsage>` 

where 

- `<submitterSubject>` is the subject from the submitter's token
- `<quotaSubject>` is the value of the HTTP `X-DataONE-QuotaSubject` header set in the request,
- `<quotaName>` is the name of the quota to check
- `requestedUsage` is the amount of the total quota to be used

 The MN will send its JWT token to authenticate this request. For storage quotas, calls to `create()` will increment the quota, calls to `update()` will increment and decrement the quota (but see the policy question below), and calls to `delete()` will decrement the quota.

> Note: @mbjones @davev - We discussed quota policies, and had questions.  Should we count all versions of objects (despite duplication) in the usage of storage quotas? Or just the `HEAD` version? An extreme case is where a `1TB` data file is updated by `1 byte` so the total usage would `2TB + 1 byte`, despite the first version being obsoleted.

If the `requestedUsage` of the `quotaName` for the `quotaSubject` does not exceed the quota hard limit, Bookkeeper will send back the `Quota` object instance as JSON.  If it exceeds it, it with throw an exception.

> Note: `MNodeService.create()` calls for *portal documents* will require two calls to `Bookkeeper.hasRemaining()` - one for the `portal_quota` and one for the `storage_quota`.
  
If it succeeds, the `create()` or `update()` can continue: 

Insert the quota details (`quota_id`, `quota_subject`, `date_reported`) in the quotas table.  Create a `QuotaUsageTask` to be submitted to the `ExecutorService` that is responsible for asynchronously updating the usage in Bookkeeper.  This task should call the Bookkeeper `adjustUsage()` method:

`PUT https://api.dataone.org/bookkeeper/v1/quotas/<quotaId>/usage?usage=<usage>`

where

- `<quotaId>` is the identifier of the quota to be updated
- `usage` is the amount to increment or decrement the total quota.  

For instance, a `create()` call for a `1MB` data file would be `usage=1048576` (or whatever the exact byte size is).  If this was an administrative call to `delete()` the file, use `usage=-1048576`.

A successful call to `adjustUsage()` will return the `Quota` object as JSON with the adjusted total `usage` field.  A failure will throw an exception.  On success, update the `quotas` table and set the `date_reported` to the current date timestamp.  On failure, set it to `NULL`.

Respond to the original `create()` or `update()`, or `delete()` request.  

> Note:  We need to think about how we manage the `quotas` table.  Multiple calls to `update()` will be likely (when someone repeatedly saves an object like a portal document or a science metadata document).  We'll need a timer process that periodically checks for `date_verified=NULL` (maybe once a night) and submits new `QuotaUsageTask`s to the `ExecutorService` so we keep Bookkeeper up to date as best as possible.

**Comment by taojing2002** [comment]
Rename the table to `usages`:
- `usage_id INT8`,  the unique usage id (pk)
- `quota_id` TEXT,  the identifier of the quota
- `instance_id` TEXT,  storage: pid of object; portal: sid of portal document
- `date_reported` TIMESTAMP,  the time stamp that the quota usage was reported to the quota service

**Comment by csjx** [comment]
Hi @taojing2002 - This looks good.  My understanding on the types for `usage_id` (`INT8`) and `quantity` (`FLOAT8`) means that we won't run out of address space even if we have millions of usages, and quantities in the terabyte range (which I guess is trillions of bytes).  For `quantity` in Bookkeeper I used a `double precision` type, which translates to `FLOAT8`.  I used `integer` for the `usageId`, but now I'm realizing I probably need to use a `serial` or `bigserial`. Since `INT8` (the same as `bigint`) includes negative numbers, you might also want to type `usage_id` as `serial` or `bigserial`.

From our conversation with Lauren and Peter, we'll be adding the `nodeId` into the `Usage` object, so while you'll need to send that along in calls to Bookkeeper, you won't need to store it here, so that part looks good too.
